### PR TITLE
feat(invoice): Refactor invoice currency handling

### PR DESCRIPTION
**Pull request by vincent-pochet**
## Context

Future changes will impact the way coupons are applied to invoices (moved before VAT computation rather than before).
In order to make this change easier, we need to improve the way we store amounts and currency on invoice.

## Description

This PR remove all `*_currency` fields from invoice in favor of a new unique `currency` field

### CODE REVIEW

File: app/models/invoice.rb [hunk 1]
```diff
@@ -22,19 +22,24 @@ class Invoice < ApplicationRecord
 
   has_one_attached :file
 
-  monetize :amount_cents
-  monetize :vat_amount_cents
-  monetize :credit_amount_cents
-  monetize :total_amount_cents
+  # NOTE: currency will be added by rails, but to enable monetize, we need to expose it before
+  def currency
```
Comment (jdenquin) [inline app/models/invoice.rb:26]: Lint/UselessMethodDefinition: Useless method definition detected.

File: app/models/invoice.rb [hunk 2]
```diff
@@ -22,19 +22,24 @@ class Invoice < ApplicationRecord
...
 
   # NOTE: Readonly fields
-  monetize :sub_total_vat_excluded_amount_cents, disable_validation: true, allow_nil: true
-  monetize :sub_total_vat_included_amount_cents, disable_validation: true, allow_nil: true
-  monetize :coupon_total_amount_cents, disable_validation: true, allow_nil: true
-  monetize :credit_note_total_amount_cents, disable_validation: true, allow_nil: true
-  monetize :charge_amount_cents, disable_validation: true, allow_nil: true
-  monetize :subscription_amount_cents, disable_validation: true, allow_nil: true
-  monetize :wallet_transaction_amount_cents, disable_validation: true, allow_nil: true
+  monetize :sub_total_vat_excluded_amount_cents, disable_validation: true, allow_nil: true, with_model_currency: :currency
```
Comment (jdenquin) [inline app/models/invoice.rb:36]: Layout/LineLength: Line is too long. [122/120] (https://rubystyle.guide#max-line-length)

File: app/models/invoice.rb [hunk 3]
```diff
@@ -22,19 +22,24 @@ class Invoice < ApplicationRecord
...
+  monetize :sub_total_vat_included_amount_cents, disable_validation: true, allow_nil: true, with_model_currency: :currency
```
Comment (jdenquin) [inline app/models/invoice.rb:37]: Layout/LineLength: Line is too long. [122/120] (https://rubystyle.guide#max-line-length)

File: db/migrate/20230414130437_add_currency_to_invoices.rb [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+class AddCurrencyToInvoices < ActiveRecord::Migration[7.0]
+  def up
+    add_column :invoices, :currency, :string
+
+    execute <<-SQL
+      UPDATE invoices
+      SET currency = amount_currency
+    SQL
+
+    remove_column :invoices, :amount_currency
```
Comment (jdenquin) [inline db/migrate/20230414130437_add_currency_to_invoices.rb:12]: Rails/BulkChangeTable: You can use `change_table :invoices, bulk: true` to combine alter queries. (https://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/SchemaStatements.html#method-i-change_table, https://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/Table.html)

File: db/migrate/20230414130437_add_currency_to_invoices.rb [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    remove_column :invoices, :vat_amount_currency
+    remove_column :invoices, :total_amount_currency
+    remove_column :invoices, :credit_amount_currency
+  end
+
+  def down
+    add_column :invoices, :amount_currency, :string
```
Comment (jdenquin) [inline db/migrate/20230414130437_add_currency_to_invoices.rb:19]: Rails/BulkChangeTable: You can use `change_table :invoices, bulk: true` to combine alter queries. (https://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/SchemaStatements.html#method-i-change_table, https://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/Table.html)

File: db/schema.rb [hunk 6]
```diff
@@ -393,16 +390,16 @@
...
     t.string "timezone", default: "UTC", null: false
     t.integer "payment_attempts", default: 0, null: false
     t.boolean "ready_for_payment_processing", default: true, null: false
     t.uuid "organization_id", null: false
     t.integer "version_number", default: 2, null: false
+    t.string "currency"
     t.bigint "fees_amount_cents", default: 0, null: false
-    t.bigint "credit_notes_amount_cents", default: 0, null: false
     t.bigint "coupons_amount_cents", default: 0, null: false
+    t.bigint "credit_notes_amount_cents", default: 0, null: false
```
Comment (rsempe) [inline db/schema.rb:402]: Seems not related to that PR

**Review by rsempe — APPROVED** [review]

File: db/schema.rb (hunk 6 above)
Comment (vincent-pochet) [reply db/schema.rb:402]: Well, I worked a lot with migration, and the order was a bit mixed :)
But if a drop create and remigrate my DB this is what I get so it will be fixed by merging one of the two PR having this issue.
